core/lib/Thelia/Action/Product.php

```diff
@@ -100,6 +121,451 @@ public function create(ProductCreateEvent $event)
+
+            $cpImageIdToLinkToPSE = ProductImageQuery::create()
+                ->filterByProductId($cpId)
+                ->findOneByPosition($opImagePosition)
+                ->getId();
+
+            $assoc = new ProductSaleElementsProductImage();
+            $assoc
+                ->setProductSaleElementsId($cpPSEId)
+                ->setProductImageId($cpImageIdToLinkToPSE)
+                ->save();
+        }
+    }
+
+    public function setClonePSEAssociatedDocuments($cpId, $cpPSEId, $opPSEDocuments)
+    {
+        foreach ($opPSEDocuments as $opPSEDocument) {
+            $opDocumentPosition = ProductDocumentQuery::create()
+                ->findOneById($opPSEDocument->getProductDocumentId())
+                ->getPosition();
+
+            $cpDocumentIdToLinkToPSE = ProductDocumentQuery::create()
+                ->filterByProductId($cpId)
+                ->findOneByPosition($opDocumentPosition)
+                ->getId();
+
+            $assoc = new ProductSaleElementsProductDocument();
+            $assoc
+                ->setProductSaleElementsId($cpPSEId)
+                ->setProductDocumentId($cpDocumentIdToLinkToPSE)
+                ->save();
+        }
+    }
+
+    public function setCloneFeatureCombination($opId, $cpId, EventDispatcherInterface $dispatcher)
+    {
+        // Get original product features
+        $opFeatures = FeatureProductQuery::create()
+            ->findByProductId($opId);
+
+        // Set clone product features
+        foreach ($opFeatures as $opFeature) {
+            $cpCreateFeatureEvent = new FeatureProductUpdateEvent($cpId, $opFeature->getFeatureId(), $opFeature->getFeatureAvId());
+            $dispatcher->dispatch(TheliaEvents::PRODUCT_FEATURE_UPDATE_VALUE, $cpCreateFeatureEvent);
+        }
+    }
+
+    public function setCloneProductAssociatedContent($opId, ProductModel $cp, EventDispatcherInterface $dispatcher)
+    {
+        // Get original product associated contents
+        $opAssocConts = ProductAssociatedContentQuery::create()
+            ->findByProductId($opId);
+
+        // Set clone product associated contents
+        foreach ($opAssocConts as $opAssocCont) {
+            $cpCreatePAC = new ProductAddContentEvent($cp, $opAssocCont->getContentId());
+            $dispatcher->dispatch(TheliaEvents::PRODUCT_ADD_CONTENT, $cpCreatePAC);
+        }
+    }
+
+    public function setCloneI18n($lang, ProductModel $op, ProductModel $cp, EventDispatcherInterface $dispatcher)
+    {
+        // Get original product's I18ns
+        $opI18ns = ProductI18nQuery::create()
+            ->findById($op->getId());
+
+        $i = 1;
+
+        foreach ($opI18ns as $opI18n) {
+            $opSeoUrl = RewritingUrlQuery::create()
+                ->filterByViewId($op->getId())
+                ->findOneByViewLocale($opI18n->getLocale())
+                ->getUrl();
+
+            $SeoRewritings = RewritingUrlQuery::create()
+                ->find();
+
```

Comment:
Selecting ALL rewritten URLs is pretty inefficient. Filter you request with $opSeoUrl